app/src/main/java/com/google/samples/apps/sunflower/compose/plantdetail/PlantDetailView.kt

```diff
@@ -0,0 +1,430 @@
+import androidx.ui.viewmodel.viewModel
+import com.google.samples.apps.sunflower.R
+import com.google.samples.apps.sunflower.compose.utils.getQuantityString
+import com.google.samples.apps.sunflower.compose.systemBarsPadding
+import com.google.samples.apps.sunflower.compose.utils.TextSnackbarContainer
+import com.google.samples.apps.sunflower.compose.visible
+import com.google.samples.apps.sunflower.data.Plant
+import com.google.samples.apps.sunflower.utilities.InjectorUtils
+import com.google.samples.apps.sunflower.viewmodels.PlantDetailViewModel
+import dev.chrisbanes.accompanist.coil.CoilImageWithCrossfade
+import dev.chrisbanes.accompanist.mdctheme.MaterialThemeFromMdcTheme
+
+
+/**
+ * As these callbacks are passed in through multiple Composables, to avoid having to name
+ * parameters to not mix them up, they're aggregated in this class.
+ */
+data class PlantDetailsCallbacks(
+    val onFabClick: () -> Unit,
+    val onBackClick: () -> Unit,
+    val onShareClick: () -> Unit
+)
+
+@Composable
+fun PlantDetailsScreen(
+    plantId: String,
+    onBackClick: () -> Unit,
+    onShareClick: (String) -> Unit
+) {
+    // ViewModel and LiveDatas needed to populate the plant details info on the screen
+    val plantDetailsViewModel: PlantDetailViewModel = viewModel(
+        factory = InjectorUtils.providePlantDetailViewModelFactory(ContextAmbient.current, plantId)
+    )
+    val plant = plantDetailsViewModel.plant.observeAsState().value
+    val isPlanted = plantDetailsViewModel.isPlanted.observeAsState().value
+    val showSnackbar = plantDetailsViewModel.showSnackbar.observeAsState().value
+
+    if (plant != null && isPlanted != null && showSnackbar != null) {
+        Surface {
+            TextSnackbarContainer(
+                snackbarText = stringResource(R.string.added_plant_to_garden),
+                showSnackbar = showSnackbar,
+                onDismissSnackbar = { plantDetailsViewModel.dismissSnackbar() }
+            ) {
+                val context = ContextAmbient.current
+                PlantDetails(plant, isPlanted, PlantDetailsCallbacks(
+                    onBackClick = onBackClick,
+                    onFabClick = {
+                        plantDetailsViewModel.addPlantToGarden()
+                    },
+                    onShareClick = {
+                        val shareText = context.resources.getString(R.string.share_text_plant, plant.name)
+                        onShareClick(shareText)
+                    }
+                ))
+            }
+        }
+    }
+}
+
+@VisibleForTesting
+@Composable
+fun PlantDetails(
+    plant: Plant,
+    isPlanted: Boolean,
+    callbacks: PlantDetailsCallbacks,
+    modifier: Modifier = Modifier
+) {
+    // PlantDetails owns the scrollerPosition to simulate CollapsingToolbarLayout's behavior
+    val scrollerPosition = ScrollerPosition()
+    var plantScroller by state {
+        PlantDetailsScroller(scrollerPosition, Float.MIN_VALUE, DensityAmbient.current)
```

Comment:
Now that I see the code, this accidentally captures the first Density seen 🤔 . It would miss changes to density (e.g. multiple screens). 

I *think* looking at this code, I'd instead pass the density to a function that returns the toolbarState (instead of making it a regular getter). WDYT?

Reply:
Yes! Changed to be passed in the function! Thanks